File: python/unified_webhook_router/utils/crypto.py
import hmac
import hashlib
import time
from typing import Union


def compute_hmac(algorithm: str, secret: str, message: Union[str, bytes]) -> str:
    if isinstance(message, str):
        message = message.encode('utf-8')
    
    if isinstance(secret, str):
        secret = secret.encode('utf-8')
    
    hash_func = getattr(hashlib, algorithm)
    return hmac.new(secret, message, hash_func).hexdigest()


def compute_hmac_base64(algorithm: str, secret: str, message: Union[str, bytes]) -> str:
    import base64
    
    if isinstance(message, str):
        message = message.encode('utf-8')
    
    if isinstance(secret, str):
        secret = secret.encode('utf-8')
    
    hash_func = getattr(hashlib, algorithm)
    signature = hmac.new(secret, message, hash_func).digest()
    return base64.b64encode(signature).decode('utf-8')


def timing_safe_compare(a: str, b: str) -> bool:
    return hmac.compare_digest(a, b)


def is_within_tolerance(timestamp: int, tolerance_seconds: int) -> bool:
    now = int(time.time())
    diff = abs(now - timestamp)
    return diff <= tolerance_seconds

File: typescript/src/utils/crypto.ts
import * as crypto from 'crypto';

export function computeHMAC(
  algorithm: string,
  secret: string,
  message: string | Buffer
): string {
  return crypto
    .createHmac(algorithm, secret)
    .update(message)
    .digest('hex');
}

export function computeHMACBase64(
  algorithm: string,
  secret: string,
  message: string | Buffer
): string {
  return crypto
    .createHmac(algorithm, secret)
    .update(message)
    .digest('base64');
}

export function timingSafeCompare(a: string, b: string): boolean {
  if (a.length !== b.length) {
    return false;
  }
  
  const bufferA = Buffer.from(a);
  const bufferB = Buffer.from(b);
  
  return crypto.timingSafeEqual(bufferA, bufferB);
}

export function isWithinTolerance(timestamp: number, toleranceSeconds: number): boolean {
  const now = Math.floor(Date.now() / 1000);
  const diff = Math.abs(now - timestamp);
  return diff <= toleranceSeconds;
}